Implement hard delete user

yishuotang opened the pull request:
**Description**

_Previous updates (2024/12):_

UI updates:
- Delete User button moved from user management main page to modify user page.
- Added first and secondary deletion confirmation modals, ensuring intentional deletion. Displaying impact details and require username confirmation, reset user input on modal close.
- Deletion only happens when both username and delete option are provided.

Schema updates:
- Added/changed a few onDelete behavior rules to ensure proper cascading deletions and nullifications.

API updates:
- Implemented hardDeleteUser function in the service layer, leverage schema rules for efficient user deletion.
- New endpoint: Added and routed **`DELETE /users/:username/delete`** to the function for hard deletion.

_Previous updates (2025/1):_

UI updates:
- "Soft Deletion" is shown and selected by default.
- "Soft Deletion" option is hidden if the user is already soft-deleted.
- Error messages are inline msgs instead of toast notifications.
- Refactored delete options to use a boolean (isHardDelete) for cleaner logic, updated deleteUser function accordingly.

API updates:
- Unified DELETE endpoint (DELETE /users/:username) for both soft and hard deletes.
-- Soft delete: DELETE /users/:username.
-- Hard delete: DELETE /users/:username?hard_delete=true.
- Unified service layer function with an isHardDelete parameter.

_New updates:_

UI Updates:

- Refactored radio button group to use `<VaRadio>` instead of `<input>`, updated related logic accordingly.
- Hard delete is now selected by default when the user is already soft-deleted.
- Moved styles into the `<style>` block.
- Displayed info message for disabled soft-deletion option using Vue’s list component inside the alert component.
- Ran linting on affected files.

API & Schema Updates:

- isHardDelete changed from string to boolean for cleaner logic.
- Implemented express-validator for request validation.
- Used a single interactive Prisma transaction for consistency.
- Updated services/user.js to provide query parameters as an object in api.delete.

**Related Issue(s)**
Closes #45 

**Changes Made**
- [x] Feature added
- [ ] Bug fixed
- [ ] Code refactored
- [ ] Documentation updated
- [ ] Other changes: [describe]

**Screenshots (if applicable)**
![clip1](https://github.com/user-attachments/assets/032422e9-c72c-4f19-88cd-5674cd42a563)
![clip2](https://github.com/user-attachments/assets/c61c7ea4-de1c-4bc9-887d-05ab7c0b8833)

**Checklist**

Before submitting this PR, please make sure that:

- [x] Your code passes linting and coding style checks.
- [ ] Documentation has been updated to reflect the changes.
- [x] You have reviewed your own code and resolved any merge conflicts.
- [x] You have requested a review from at least one team member.
- [x] Any relevant issue(s) have been linked to this PR.

deepakduggirala commented:
UX Design Recommendation:

The button to hard delete user should not be readily available. My advise is to move the button from the table into the user details page. Also, ask the user to type in the username they are trying to delete to absolutely confirm that is what they want to do as this process is irreversible.

deepakduggirala commented:
DB Delete Cascade:

In the database, user table has a lot of relations. When a user record is deleted one of these three things happen:
- The associated relation record is deleted
- The associated relation record is updated to set user_id to null
- If unspecified, the delete op fails

yishuotang commented:
Hi @deepakduggirala and @ri-pandey, I’ve made updates based on your feedback, thank you so much for the suggestions!
I’ve updated the PR description with more details and included GIF's demonstrating the feature. Please let me know if there’s anything else to address!
(The branches delete-user-feature appear in the commit graph likely due to shared history, but this PR is specifically tied to the original branch 45---Implement-hard-delete-user. Please review the changes in this branch only. :) )

ri-pandey commented on ui/src/pages/users.vue:
```diff
@@ -468,6 +582,47 @@ function createUser() {
 
+function hardDeleteUser() {
+  if (usernameConfirmation.value !== editedUser.value.username) {
+    toast.error('Username does not match. Entered: "${usernameConfirmation.value}", Expected: "${editedUser.value.username}". ');
```
Instead of showing this error via toast, show use `<va-input>`'s props to show error directly beneath the input. See examples of how to achieve this here: https://ui.vuestic.dev/ui-elements/input#validation-and-error-success-state

ri-pandey replied:
I guess this has been replaced with the error message "Username mismatch detected during deletion attempt.". But I am not sure when that error message should be visible on the UI. You could either post a GIF, or we can review this on Zoom.

yishuotang replied:
I didn’t generate an inline message for this use case because if the username isn’t typed correctly, the confirm button on the bottom right remains disabled, so I suppose the error msg wouldn't show up in practice. Do you think it would be helpful to include it as a safety net? Happy to adjust if needed! :)

ri-pandey replied:
I think it would be helpful to see the error so the user knows why the Submit is disabled. Something like "Username does not match expected username... ".

ri-pandey commented on ui/src/pages/users.vue:
```diff
@@ -468,6 +582,47 @@ function createUser() {
+
+  if (!deleteOption.value) {
+    toast.error("Please select an action for deletion (hard or soft deletion).");
+    return;
```
Show errors beneath input instead of toast.

ri-pandey replied:
The Soft/Delete choice is a radio button group, and by default, one radio button in a group should always be selected. So when do we expect to see this error (the updated error that I was trying to see in action is "Please select a valid delete option.")?

yishuotang replied:
This error message appears when a username is already marked as soft_deleted, and an admin opens the confirmation modal, types in the correct username, but clicks the confirm button without selecting the 'hard deletion' option. Since 'soft delete' is selected by default, this error only occurs if no valid delete option is selected in this scenario. (as shown in gif #2)
```
// Prevent soft delete if the user is already soft deleted/ no options selected
  if (!isHardDelete.value && isSoftDeleted.value) {
    deleteOptionError.value = "Please select a valid delete option.";
    console.warn("Soft delete attempted on an already soft-deleted user.");
    return;
  }
```

ri-pandey replied:
Can you replace your usages of `<input>` to `<va-input />`? The rest of the app uses Vuestic's components, so we should be doing the same here.

Also, providing both `v-model` and `value` is redundant. You should only need `v-model`, at least once you switch to Vuestic's <va-input /> component.

ri-pandey replied:
@yishuotang If the user is already soft-deleted, I would handle that case like this:

1. Keep the 'soft-delete' radio-button disabled, as you are currently doing
2. Make the 'hard-delete' option selected by default. This way, there won't be a scenario where you need to show the "Please select a valid delete option" message.
3. Instead of showing the "soft-delete is disabled because..." message as an error beneath the `<va-input />`, show this message through a Vuestic Alert (https://ui.vuestic.dev/ui-elements/alert). The error message beneath the input makes more sense for form validation errors, like "username does not match...", not for informational messages.

ri-pandey commented on api/src/routes/users.js:
```diff
@@ -131,4 +132,19 @@ router.delete(
+// add new route to hard delete a user
+router.delete(
+  '/:username/delete',
+  isPermittedTo('delete', { checkOwnerShip: true }),
```
Can you use a single DELETE endpoint for both hard and soft-delete cases? Try to reuse the existing one.

To reuse, send a query parameter called `hard_delete` in the URL (in the existing DELETE endpoint). Default it to `false`. Send the query param as `true` if you want to hard-delete. Then, within the endpoint definition, check the value of `req.query.hard_delete` to determine if the endpoint needs to do a soft-deletion or a hard-deletion.

Examples of boolean query params: https://github.com/IUSCA/bioloop/blob/main/api/src/routes/datasets.js#L284

ri-pandey replied:
Also, please ensure (if you haven't already) that any new routes you add are idempotent. That is, the end result of the operation (for example, database updates done by the route) stays the same, no matter how many times the endpoint is called.

ri-pandey replied:
Currently, line 31 is `const hardDelete = req.query.hard_delete === 'true';`.

It's better to interpret true/false as boolean values on the API side, instead of strings.

If your API request URL looks something like this: `https://bioloop-dev.sca.iu.edu/api/users/ripandey?hard_delete=true`

then you can get the Express-validator library (already in the project) to validate any query parameters like `hard_delete`, and also convert them from strings to booleans on the fly.

Here is an example of how to do that: https://github.com/IUSCA/bioloop/blob/main/api/src/routes/datasets.js#L205. This example also defaults the `deleted` query parameter to `false`. You could do the same thing for the `hard_delete` parameter, so that if the API request doesn't provide this parameter, it defaults to `false`.

ri-pandey commented on api/src/services/user.js:
```diff
@@ -162,6 +162,37 @@ async function softDeleteUser(username) {
+    const userDetails = { id: user.id, username: user.username };
+    // Delete the user (cascading delete and nullifications are handled by the schema)
+    await prisma.user.delete({ where: { id: user.id } });
+
```
Try to use a single Prisma transaction. This can be done thus:

```
await prisma.delete(
  { where: username: 'someUsername' }
)
```

The `username` field has a `@unique` constraint, so using prisma's `delete` method will find and delete the record corresponding to that username only.

ri-pandey replied:
It is still possible to do this within a transaction, in the case of soft-deleting a user. You can use Prisma's interactive transaction API to do this - https://www.prisma.io/docs/orm/prisma-client/queries/transactions#interactive-transactions.

I would be happy to guide you through how interactive transactions work, if it saves you some time.

ri-pandey commented on ui/src/pages/users.vue:
```diff
@@ -236,6 +251,58 @@
+      </label>
+      <label class="ml-4">
+        <input type="checkbox" :checked="deleteOption === 'softDelete'" @change="toggleOption('softDelete')" />
+        Soft Deletion
```
Could you not show the `Soft Deletion` radio button option if the user is already soft-deleted? Instead, just show `Hard Deletion` and make it selected by default

yishuotang commented:
Hi @ri-pandey , thanks again for your suggestions. I’ve implemented the changes and updated the PR. Let me know if there’s anything else I should address!

ri-pandey commented:
Can you run lint on the affected code within ui/ and api/?

Something like:
```
cd ui
npm install
npm run lint
# This will fix the linting in all UI files. It's best to only commit the linted changes to files that were affected in this ticket.
```

Then, run the same steps within the api directory.

ri-pandey commented on ui/src/pages/users.vue:
```diff
@@ -262,6 +342,61 @@ const autofill = ref({
+const usernameConfirmation = ref("");
+//const deleteOption = ref(null);
+const isHardDelete = ref(false);
+const impactDetails = ref(null);
```
If you are setting `impactDetails` to an array, then it makes sense to also initialize it as an array, like this:

`const impactDetails = ref([])`

Also, if you initialize this as an array, all the places where you are checking for `impactDetails` to be `null` will need to be replaced with `impactDetails.length > 0` (within the `<template>` blocks) or `impactDetails.value.length > 0` (within the `<script />` blocks).

And the places where you are setting `impactDetails` to null will need to be changed, so that it is set to `[]` instead.

ri-pandey commented on ui/src/services/user.js:
```diff
@@ -27,10 +27,34 @@ class UserService {
+
+  // Unified deleteUser function
+  deleteUser(username, isHardDelete = false) {
+    const queryParam = isHardDelete ? "?hard_delete=true" : ""; 
```
The query parameters can probably be provided as an object argument to the `api.delete` method, and shouldn't need to be set as a string like `?hard_delete=true`.

ri-pandey commented:
Two more suggestions:

1. Once you have replaced your usages of `<input>` with `<va-input>`, make sure that you are using va-input's `error` and `error-messages` props to show errors. This way, you won't have to write custom markup like:
```
<div v-if="deleteOptionError" class="text-danger mt-1" style="color: red; font-size: 0.875rem;">
      {{ deleteOptionError }}
    </div>
```
to show errors beneath the input. You can find examples at https://ui.vuestic.dev/ui-elements/input.

2. I haven't inspected all the usages of inline styles (i.e. using `style="..."` within the `<template>` blocks). I did comment on one usage, but if there are more, it's better to move them to <style> blocks, so that those styles can be easily overridden in the future, which is difficult to do with inline styles.

3. This last point is just a suggestion, I would be happy to approve for now even without this: it's better to use Vuestic's Alert component to show information messages like "the following user data will be affected by soft/hard-delete...". For showing multi-line alerts, you can use Vuestic's List component (https://ui.vuestic.dev/ui-elements/list) within Vuestic's Alert component.

yishuotang commented:
Hi @ri-pandey , thank you very much for the detailed suggestions! I’ve implemented the changes and updated the PR. Please check out the new GIFs for a quick overview of the UI. Let me know if there’s anything else I should address!

ri-pandey commented on ui/src/pages/users.vue:
```diff
@@ -512,6 +705,23 @@ fetch_all_users();
+.deletion-options {
+  display: flex;
+  gap: 10px; /* Space between radio buttons */
+}
```
These two CSS styles can be applied through the `class` prop. I think something like `class="flex gap-10"` should do it. Tailwind flex/gap docs - https://tailwindcss.com/docs/flex and  https://tailwindcss.com/docs/gap

ri-pandey commented on ui/src/pages/users.vue:
```diff
@@ -262,6 +364,56 @@ const autofill = ref({
+  isDeleteModalVisible.value = true;
+}
+
+const getImpactDetails = computed(() => {
```
It's better to avoid naming computed properties like getters. So, let's rename this to `impactDetails` if possible, or some other name that does not contain the word `get`.
